packages/@lwc/template-compiler/src/__tests__/fixtures/directive-inner-html/valid/expected.js

```diff
@@ -0,0 +1,38 @@
+import { registerTemplate, sanitizeHtmlContent } from "lwc";
+function tmpl($api, $cmp, $slotset, $ctx) {
+  const { h: api_element } = $api;
+  return [
+    api_element(
+      "div",
+      {
+        props: {
+          innerHTML: sanitizeHtmlContent("Hello <b>world</b>!"),
```

Comment:
Now that I am looking at this code again I think that it should be optimized. Currently, every time the template is evaluated, `sanitizeHtmlContent` is invoked. With this PR, the LWC engine diffs the sanitized strings and not the original strings. So if the HTML string is identical between 2 render cycles it still has to be sanitized again.

This might not be an issue for small HTML content, but DOMPurify can take more than 100ms to sanitize 1MB of raw HTML ([benchmarks](https://jsfiddle.net/5askurp0/12/)). One approach would be to do something similar as [event listeners caching](https://github.com/salesforce/lwc/blob/master/packages/@lwc/template-compiler/src/__tests__/fixtures/events/tag/expected.js#L17) to avoid wasting cycles sanitizing the same content between rendering cycles.

FYI @jodarove

Reply:
@pmdartus just double-checking before implementing. I totally agree that `sanitizeHtmlContent` is expensive, however, this PR is leveraging into the diffing algo and existing runtime, therefore `<value>` in `props: { innerHTML: <value>` must be the sanitized string, because I don't think we should modify the runtime to handle the `innerHTML` since we will be punishing updating rest of the props.

What you suggest is to create a new API `api_sanitize_str`, that returns a function that caches the previous (value, sanitizedValue) to avoid calling `sanitizeHtmlContent` if the previous value and the new one are the same? For example:

```js
// compiled template
// ...
const { _s0, _s1 } = $ctx; // as many as lwc:innerHtml are used in the component template.

api_element(
      "div",
      {
        props: {
          innerHTML: (_s0 || ($ctx._s0 = api_sanitize_str()))("Hello <b>world</b>!"),
```

```js
function api_sanitize_str() {
   let currentValue = '';
   let sanitizedValue = '';

   return function(newValue) {
      if (value !== previousValue) {
          currentValue = newValue;
          sanitizedValue = sanitize_html(newValue);
      }

      return sanitizedValue;
   }
}
```

Reply:
Yes, something similar to this. Keep in mind that you might not know all the `lwc:inner-html` ahead of time, if the directive is used inside a `for:each` block.

I was wondering if it is possible to get rid of the closure. And yes you can do it by storing both the raw value and the sanitized content on the `$ctx` object. That said it creates a really ugly expression. I am fine with going the closure route since the overhead will be pretty minimal.

```html
<template>
    <div lwc:inner-html={value}></div>
</template>
```

```js
function tmpl($api, $cmp, $slotset, $ctx) {
  const { h: api_element } = $api;

  return [
    api_element(
      "div",
      {
        props: {
          innerHtml:
            $ctx._rawHtml$1 !== ($ctx._rawHtml$1 = $cmp.value)
              ? ($ctx._sanitizedHtml$1 = sanitizeHtml($ctx._rawHtml$1))
              : $ctx._sanitizedHtml$1,
        },
        key: 0,
      },
      []
    ),
  ];
}
```